### PR TITLE
Add executors E2E pipeline

### PR DESCRIPTION
**Pull request by jhchabran**
Edit by @eseliger 

---

This PR adds the infrastructure to run and an initial test for server-side batch changes with executors.

I've implemented a basic testing framework so that we can in the future stuff in a few more tests here to test all the features of batch changes in this pipeline, this will become increasingly useful as we will want to transition to full native execution.

Big thanks to @jhchabran for all the help getting this pipeline to work!

## Test plan

Verified it passes locally and in CI.

### CODE REVIEW

File: dev/ci/integration/executors/config/site-config.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "executors.accessToken": "hunter2hunter2hunter2",
```
Comment (eseliger) [inline dev/ci/integration/executors/config/site-config.json:2]: TODO: From env var?

File: dev/ci/integration/executors/config/site-config.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  "auth.providers": [
+    {
+      "type": "builtin",
+      "allowSignup": false
+    }
+  ],
+  "externalURL": "http://localhost:7080/",
```
Comment (eseliger) [inline dev/ci/integration/executors/config/site-config.json:9]: TODO: From env var?

File: internal/gqltestutil/batch_specs.go [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+			path
+			name
+		  }
+		}
+	  }
+	}
+}
+`
+
+type BatchSpecDeep struct {
```
Comment (eseliger) [inline internal/gqltestutil/batch_specs.go:317]: Would be great if we had a codegen-ed type here from the query above.

File: enterprise/dev/ci/internal/ci/operations.go [hunk 4]
```diff
@@ -622,6 +622,23 @@ func codeIntelQA(candidateTag string) operations.Operation {
 	}
 }
 
+func executorsE2E(candidateTag string) operations.Operation {
```
Comment (eseliger) [inline enterprise/dev/ci/internal/ci/operations.go:625]: It seems like the server cannot start, didn't investigate further.

File: dev/ci/integration/executors/run.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+TMP_DIR="${TMP_WORK_DIR}/executor-tmp"
+mkdir "${TMP_DIR}"
+DATA="${TMP_WORK_DIR}/data"
+mkdir "${DATA}"
+mkdir "${DATA}/data"
+mkdir "${DATA}/config"
+
+cleanup() {
+  pushd "$root_dir"/dev/ci/integration/executors/ 1>/dev/null
+  docker-compose logs >logs.txt
```
Comment (eseliger) [inline dev/ci/integration/executors/run.sh:19]: TODO: Make artifact

File: dev/ci/integration/executors/tester/main.go [hunk 6]
```diff
@@ -0,0 +1,547 @@
+package main
```
Comment (eseliger) [inline dev/ci/integration/executors/tester/main.go:1]: TODO: Clean up and make more reusable for additional tests.
TODO: Assert on changeset specs and execution cache results in DB (meta: Can we read from the DB in E2E tests?)

File: dev/ci/integration/executors/tester/main.go [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+	logger.Info("Loading batch spec to assert")
+
+	batchSpec, err := client.GetBatchSpecDeep(batchSpecID)
+	if err != nil {
+		return err
+	}
+	if batchSpec.State != "COMPLETED" {
+		logger.Fatal("batch spec is not complete")
+	}
+	if diff := cmp.Diff(batchSpec, &gqltestutil.BatchSpecDeep{
```
Comment (eseliger) [inline dev/ci/integration/executors/tester/main.go:223]: TODO: This is very ugly

File: dev/ci/integration/executors/run.sh (hunk 5 above)
Comment (jhchabran) [reply dev/ci/integration/executors/run.sh]: done

File: enterprise/dev/ci/internal/ci/operations.go [hunk 8]
```diff
@@ -622,6 +622,29 @@ func codeIntelQA(candidateTag string) operations.Operation {
...
+	return func(p *bk.Pipeline) {
+		for i := 0; i < 100; i++ {
```
Comment (eseliger) [inline enterprise/dev/ci/internal/ci/operations.go:627]: TODO: Remove

File: enterprise/dev/ci/integration/executors/tester/main.go [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+  branch: hello-world # Push the commit to this branch.
+  commit:
+    message: Append Hello World to all README.md files
+`
+
+var expectedDiff = strings.Join([]string{
+	"diff --git README.md README.md",
+	"index 1914491..89e55af 100644",
+	"--- README.md",
+	"+++ README.md",
```
Comment (unknwon) [inline enterprise/dev/ci/integration/executors/tester/main.go:138]: qq: why not use

```go
const expectedDiff = `diff --git README.md README.md
index 1914491..89e55af 100644
--- README.md
+++ README.m
...`
```

?

Comment (jhchabran) [reply enterprise/dev/ci/integration/executors/tester/main.go:138]: ported from slack: 

> the reason is that git patches encode empty lines as ` ` (space) and in backtick strings gofmt removes trailing spaces on save :smile:

File: enterprise/dev/ci/integration/executors/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  executors-e2e:
+    external: true
+services:
+  # Run an external postgres so we can access it from the tester.
+  postgres:
+    container_name: postgres
+    image: '${POSTGRES_IMAGE}'
+    expose:
+      - '5432'
+    ports:
```
Comment (davejrt) [inline enterprise/dev/ci/integration/executors/docker-compose.yml:14]: Might be worth adding a comment here, as to why we use 5433 (so it doesn't conflict with other pg instances) etc

File: enterprise/dev/ci/integration/executors/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      - EXECUTOR_QUEUE_NAME=batches
+      # Note: Must match left-hand side of scratch volume mount
+      - EXECUTOR_DOCKER_HOST_MOUNT_PATH=${TMP_DIR}
+      # Note: Must match right-hand side of scratch volume mount
+      - TMPDIR=/scratch
+      # Run as root (required for docker daemon control)
+      - UID=1000
+      - GID=1000
+      # Use the dind container to run docker commands within the executor
+      - DOCKER_HOST=${DOCKER_HOST}
```
Comment (davejrt) [inline enterprise/dev/ci/integration/executors/docker-compose.yml:67]: Nice!

**Review by davejrt — COMMENTED** [review]
A couple of quick questions, but overall looks good.

EDIT: Approving based on the quick feedback, some minor inline comments etc but NICE WORK @jhchabran @eseliger

File: enterprise/dev/ci/internal/ci/operations.go [hunk 12]
```diff
@@ -622,6 +622,27 @@ func codeIntelQA(candidateTag string) operations.Operation {
...
+			bk.DependsOn(candidateImageStepKey("executor")),
+			bk.Env("CANDIDATE_VERSION", candidateTag),
+			bk.Env("SOURCEGRAPH_BASE_URL", "http://127.0.0.1:7080"),
+			bk.Env("SOURCEGRAPH_SUDO_USER", "admin"),
+			bk.Env("TEST_USER_EMAIL", "test@sourcegraph.com"),
+			bk.Env("TEST_USER_PASSWORD", "supersecurepassword"),
+			// See enterprise/dev/ci/integration/executors/docker-compose.yaml
+			// This enable the executor to reach the dind container
+			// for docker commands.
+			bk.Env("DOCKER_GATEWAY_HOST", "172.17.0.1"),
```
[the remark this answers is not in the document]
Comment (jhchabran) [reply enterprise/dev/ci/internal/ci/operations.go:639]: Yes!

[the remark this answers is not in the document]
Comment (davejrt) [reply enterprise/dev/ci/internal/ci/operations.go:639]: Got it, that IP range looked familiar but just wanted to make sure we were getting all our subnets and localhosts/docker_hosts crossed.

**Review by davejrt — APPROVED** [review]

**Review by Piszmog — APPROVED** [review]
Looks good!

**Pull request merged**